
<div style="text-align: center;">
  <img src="https://github.com/user-attachments/assets/447a923e-c4d1-4331-8a81-130f48144ca0" alt="Breakinator Logo" width="300"/>
</div>

# The Breakinator
The Breakinator identifies and flags putative artifact reads (foldbacks and chimeric) by parsing a PAF file.

## Installation

```
git clone git@github.com:jheinz27/breakinator.git
cd breakinator
python breakinator.py -h
```

#### Prerequisites

- Python 3.7 or higher  

## Generating PAF files

To generate the PAF files to input to the Breakinator, we recommend using [minimap2](https://github.com/lh3/minimap2) with the `-c` and `--secondary=no` parameters. Seecondary alignments will be ignored by the Breakinator, however including them will increase the processing time. 

Example:
```
minimap2 -cx map-ont --secondary=no genome.fa reads.fastq > alignments.paf
```

The PAF can also be generated by converting a SAM file to a PAF with [paftools.js](https://github.com/lh3/minimap2/blob/master/misc/paftools.js) using the `-p` parameter. 

Example: 
```
paftools.js sam2paf -p alignments.sam > alignments.paf
```

## Breakinator Usage
```
usage: breakinator.py [-h] -i FILE [-m INT] [-a INT] [--sym] [--rcoord] [--margin FLOAT] [-o FILE] [--chim INT] [--fold INT] [--tabular]

Flag foldbacks and chimeric reads from PAF input

optional arguments:
  -h, --help      show this help message and exit
  -i FILE         PAF file sorted by read IDs
  -m INT          Minimum mapping quality (integer)
  -a INT          Minimum alignment length (bps)
  --sym           Only report palindromic foldback reads within margin
  --rcoord        Print read coordinates of breakpoint in output
  --margin FLOAT  [0-1], With --sym, Proportion from center on either side to be considered foldback artifact
  -o FILE         Output file name
  --chim INT      Minimum distance to be considered chimeric
  --fold INT      Max distance to be considered foldback
  --tabular       Return report as a tsv file (useful for evaluating multiple files)
```

## Optional symmetry filter for foldback artifacts

If running on a sample where you expect there may be true foldback events (e.g cancer data) we recommend using the `--sym` flag to only consider reads that have the foldback withing 5% (change with `--margin`) on either side of the middle of the read as foldback artifacts. See diagram below. 

```
python breakinator.py -i alignments.paf --sym --margin 0.03
```
<img width="742" alt="Screenshot 2025-05-09 at 10 15 35â€¯AM" src="https://github.com/user-attachments/assets/c66855bb-5fbd-4143-a884-9bd200a4395f" />

## Preprocessing for alignment to diploid genome assemblies with The Diploidinator

Minimap2 was not designed for diploid assemblies(eg. [HG002](https://github.com/marbl/HG002)), so when aligning reads to a diploid assembly, the mapping quality for reads may be lower, as there are multiple locations the read can align to well. We have developed a simple rust script to align reads to each haploid of the diploid assembly and then parse both paf files to choose the better alignment of the read based on the alignment score. 

---

## Diploidinator Installation

``` 
git clone git@github.com:jheinz27/breakinator.git
cd breakinator/diploidinator
cargo build --release
./target/release/diploidinator
```

## Diploidinator Example Usage

NOTE: It is important to use the `--secondary=no` and `--paf-no-hit` flags when aligning with Minimap2. The diploidinator currently only works on paf files. 
```
minimap2 -cx splice -uf -k14 -t 16 --secondary=no --paf-no-hit hg002v1.1.MATERNAL.fasta read.fastq > out_mat.paf
minimap2 -cx splice -uf -k14 -t 16 --secondary=no --paf-no-hit hg002v1.1.PATERNAL.fasta reads.fastq > out_pat.paf 
diploidinator out_mat.paf out_pat.paf > out_haps_merge.paf
```
## Merging Breakpoints Into Consensus Locations
To evaluate how many unique breakpoints are in the sample and how much read support they have, we developed a simple script to merge breakpoints together if they occur within 100bps (default `-w`) of eachother. We require at least 2 reads (default `-s`) of support to report a consensus breakpoint location. 
```
usage: merge_breaks.py [-h] -i <breakpoints.txt> [-w --merge_window] [-s --min_support] > merged_breaks.txt

Merge Break Points from Breakinator output

optional arguments:
  -h, --help            show this help message and exit
  -i <breakpoints.txt>  input breakinator stdout
  -w --merge_window     Size of window to merge break points in
  -s --min_support      minimum reads supporting breakpoint
```


## Citation
If the Breakinator has helped you in your research, please cite our preprint at: 


